oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elasticsearch/index/ElasticsearchIndexWriter.java

```diff
@@ -142,7 +149,37 @@ public boolean close(long timestamp) throws IOException {
     // TODO: we need to check if the index already exists and in that case we have to figure out if there are
     // "breaking changes" in the index definition
     protected void provisionIndex() throws IOException {
-        CreateIndexRequest request = new CreateIndexRequest(indexDefinition.getRemoteIndexName());
+
+        IndicesClient indicesClient = elasticsearchConnection.getClient().indices();
+        final String indexName = indexDefinition.getRemoteIndexName();
+
+        CreateIndexRequest createIndexRequest = constructCreateIndexRequest(indexName);
+        String requestMsg = Strings.toString(createIndexRequest.toXContent(jsonBuilder(), EMPTY_PARAMS));
+        CreateIndexResponse response = indicesClient.create(createIndexRequest, RequestOptions.DEFAULT);
+
+        LOG.info("Updated settings for index {} = {}. Response acknowledged: {}",
+                indexDefinition.getRemoteIndexAlias(), requestMsg, response.isAcknowledged());
+
+
+        GetAliasesRequest getAliasesRequest = new GetAliasesRequest(indexDefinition.getRemoteIndexAlias());
+        GetAliasesResponse aliasesResponse = indicesClient.getAlias(getAliasesRequest, RequestOptions.DEFAULT);
+        Map<String, Set<AliasMetaData>> aliases = aliasesResponse.getAliases();
+        IndicesAliasesRequest indicesAliasesRequest = new IndicesAliasesRequest();
+        for (String oldIndexName : aliases.keySet()) {
+            IndicesAliasesRequest.AliasActions removeAction = new IndicesAliasesRequest.AliasActions(IndicesAliasesRequest.AliasActions.Type.REMOVE);
+            removeAction.index(oldIndexName).alias(indexDefinition.getRemoteIndexAlias());
+            indicesAliasesRequest.addAliasAction(removeAction);
```

Comment:
what about deleting the old indexes returned from the get alias call? There should be always one index. The delete should be the last operation we do here to be sure the alias has been swapped

Reply:
Done. I wasn't able to use the [remove_index](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-aliases.html) as the last action in update index call itself. It threw exception that index doesn't exist. From ES logs it seems deletion of old index happened before switching alias, so no particular ordering is being followed while executing actions - 
```
[2020-05-11T14:35:45,825][INFO ][o.e.c.m.MetaDataCreateIndexService] [wx-3] [aem4may._eindexone-2] creating index, cause [api], templates [], shards [1]/[1], mappings [_doc]
[2020-05-11T14:35:46,554][INFO ][o.e.c.m.MetaDataDeleteIndexService] [wx-3] [aem4may._eindexone-1/NysOwkcLT8OZDMvq3cQ7IA] deleting index
[2020-05-11T14:35:46,563][DEBUG][o.e.a.s.m.TransportMasterNodeAction] [wx-3] failed to perform aliases
org.elasticsearch.index.IndexNotFoundException: no such index [aem4may._eindexone-1]
        at org.elasticsearch.cluster.metadata.MetaDataIndexAliasesService.innerExecute(MetaDataIndexAliasesService.java:126) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.cluster.metadata.MetaDataIndexAliasesService$1.execute(MetaDataIndexAliasesService.java:88) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.cluster.ClusterStateUpdateTask.execute(ClusterStateUpdateTask.java:47) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.cluster.service.MasterService.executeTasks(MasterService.java:702) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.cluster.service.MasterService.calculateTaskOutputs(MasterService.java:324) ~[elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.cluster.service.MasterService.runTasks(MasterService.java:219) [elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.cluster.service.MasterService.access$000(MasterService.java:73) [elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.cluster.service.MasterService$Batcher.run(MasterService.java:151) [elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.cluster.service.TaskBatcher.runIfNotProcessed(TaskBatcher.java:150) [elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.cluster.service.TaskBatcher$BatchedTask.run(TaskBatcher.java:188) [elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:703) [elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:252) [elasticsearch-7.5.1.jar:7.5.1]
        at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:215) [elasticsearch-7.5.1.jar:7.5.1]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:830) [?:?]
```

So I deleted the index in a separate Request after switching the alias.